// react-day-picker/build/contexts/DayPicker/utils/parseModifierProps.js
/** Props that will merge into the modifiers. */
var modifierShortcuts = [
    'selected',
    'hidden',
    'disabled'
];
/**
 * Parse the modifiers from the props.
 *
 * Internally we want modifiers as an array of matchers – as opposite of the
 * props which can accept also a matcher.
 */
function parseModifierProps(initialProps) {
    var initialModifiers = initialProps.modifiers || {};
    var modifiers = {
        selected: [],
        disabled: [],
        hidden: [],
        today: [],
        range_end: [],
        range_middle: [],
        range_start: []
    };
    Object.entries(initialModifiers).forEach(function (_a) {
        var modifier = _a[0], matcher = _a[1];
        if (Array.isArray(matcher)) {
            modifiers[modifier] = matcher;
        }
        else if (matcher) {
            modifiers[modifier] = [matcher];
        }
        else {
            modifiers[modifier] = [];
        }
    });
    modifierShortcuts.forEach(function (modifier) {
        // @ts-expect-error TOFIX: initial props do not have some modifiers
        var modifierFromProp = initialProps[modifier];
        if (Array.isArray(modifierFromProp)) {
            modifiers[modifier] = modifierFromProp;
        }
        else if (modifierFromProp !== undefined) {
            modifiers[modifier] = [modifierFromProp];
        }
        else {
            modifiers[modifier] = [];
        }
    });
    return modifiers;
}

export { parseModifierProps };
